### PR TITLE
Generate RZIL operations

### PR DESCRIPTION
**Pull request by Rot127**
## WIP

Requires: https://github.com/rizinorg/rz-hexagon/pull/67

The `rzil-compile` sub-module used here is at https://github.com/Rot127/rzil-compiler

closes https://github.com/rizinorg/rz-hexagon/issues/64

### CODE REVIEW

File: LLVMImporter.py [hunk 1]
```diff
@@ -442,11 +622,67 @@ def build_hexagon_disas_c(self, path: str = "./rizin/librz/asm/arch/hexagon/hexa
...
+        code += include_file("handwritten/hexagon_il_getter_table_h/includes.h")
+        code += "\n"
+
+        # Lookup table
+        code += "static HexILInsn hex_il_getter_lt[] = {\n"
+        table = ""
+        for name in sorted(self.normal_instruction_names + self.sub_instruction_names):
+            insn = self.normal_instructions[name] if name in self.normal_instructions else self.sub_instructions[name]
+            if "invalid_decode" in name.lower():
+                # Invlaid decode is always at the top.
```
Comment (DMaroo) [inline LLVMImporter.py:642]: Typo: ~invlaid~ -> invalid

**Review by DMaroo — COMMENTED** [review]
Glanced through the handwritten code, and it looks pretty good to me. Didn't really have the time to thoroughly read through it. If you have any files which you would want me to review throughly, I'd be happy to do so. Great stuff, btw!

**Comment by Rot127** [comment]
Looking at the Python files is fine. No need to read too deep into it, if you haven't already.
The hand-written C code can be reviewed in the Rizin PR IMHO.

@DMaroo More importantly I would need feedback on the compiler (https://github.com/Rot127/rzil-compiler). Though only if you find time and are interested for using it in the future. Do not want to waste your time.

I designed it roughly in a way that we should be able to use it for other archs as well. "Roughly" means I only used it for Hexagon yet. And have not tested it yet with another arch.
Also it is not a good compiler. Its pretty naively implemented and slow. But does it's job.

**Comment by DMaroo** [comment]
Got it, thanks! I'll have a look at https://github.com/Rot127/rzil-compiler once I have time.

**Comment by thestr4ng3r** [comment]
Do you think it would be possible to further reduce the code size? For example `rizin/librz/analysis/arch/hexagon/il_ops/hexagon_il_A2_ops.c` has 12k lines of code, which is absolutely massive. For example by moving more logic from the rzil compiler/code generator to the runtime.

**Comment by Rot127** [comment]
Yes. Already did it but haven't pushed yet. `A2_ops` is down to ~7700 lines. And it can be reduced a little further.

**Comment by Rot127** [comment]
So for `A2` the line count is now at roughly 8k. But this is currently as good as it gets (of cause everything but the effects could be inline, but this is not really the point).

To tackle instructions like [these](https://github.com/rizinorg/rizin/blob/3f283a4c8e0cae29fc890e192000ed4d0174ccab/librz/analysis/arch/hexagon/il_ops/hexagon_il_M2_ops.c#L125) sub-routines need to be implemented.

Its not too difficult but needs time. So before that I'd like to get the currently 1500 instruction tested.

**Comment by XVilka** [comment]
I was thinking about supporting subroutines/functions in RzIL for a long time too. It will help to simplify some of the output, will also make easier spotting stuff like input/output and emulating those, if necessary.

**Comment by Rot127** [comment]
Couldn't find a corresponding concept in BAPs IL. Hopefully I haven't just overlooked it.

But for now I had to implemented something which could be used for them.
Behavior of C syntax which has to be categorized as `Effect` and `Pure` (think of `i++` but also sub-routines), get generated like this:

```c
Effect *sub_routine_sequence = sub_routine(...); // Sets VARL("return_val_i") in its sequence
Pure *return_val = CAST(ret_width, ret_sign, VARL("return_val_i"))
```

Not smart, but this is how I would do those sub-routines for now.

**Pull request merged**